### PR TITLE
Migrate k8s to kubernetes/client-go

### LINKED ISSUE #532: Cannot connect to Kubernetes server failed to decode watch event
Hi,

I was using Traefik v1.0.0 with Kubernetes 1.3.0. Randomly I'm getting invalid character errors and it crashes Traefik after a few failures. Any one experiencing the same problem?

```
time="2016-07-14T10:58:09Z" level=fatal msg="Cannot connect to Kubernetes server failed to decode watch event: GET \"https://10.254.0.1:443/api/v1/endpoints?watch&resourceVersion=2928703\" : invalid character '\\n' in string literal" 
time="2016-07-14T10:19:06Z" level=fatal msg="Cannot connect to Kubernetes server failed to decode watch event: GET \"https://10.254.0.1:443/api/v1/endpoints?watch&resourceVersion=2924878\" : invalid character 'a' looking for beginning of value" 
time="2016-07-14T09:00:01Z" level=fatal msg="Cannot connect to Kubernetes server failed to decode watch event: GET \"https://10.254.0.1:443/api/v1/endpoints?watch&resourceVersion=2916459\" : invalid character 'o' looking for beginning of value" 
time="2016-07-14T08:15:02Z" level=fatal msg="Cannot connect to Kubernetes server failed to decode watch event: GET \"https://10.254.0.1:443/api/v1/endpoints?watch&resourceVersion=2911726\" : invalid character '}' looking for beginning of value" 
```

Thanks!

### PR DESCRIPTION
**Pull request by yvespp**
Implements #678 should fix #532
This also dramatically reduces the amount of k8s API requests traefik makes. The client uses watches to update local caches that are then used to build the traefik config.

Outstanding Issues:
1. glog flag redefined error: http://stackoverflow.com/questions/37284423/glog-flag-redefined-error
The k8s client vendors glog which conflicts with the glog from another import. I solved this by deleting the glog folder in scripts/generate.sh: `rm -rf vendor/k8s.io/client-go/1.5/vendor/github.com/golang/glog/`
How to solve this better?
2. I removed the config option `Endpoint` for the k8s provider. This breaks backward compatibility. I think it would be better to support the kubectl config file for an out of cluster config. The client supports this easily: https://github.com/kubernetes/client-go/blob/release-1.5/examples/out-of-cluster/main.go
3. The client logs this errors from time to time:
`ERROR: logging before flag.Parse: W1111 20:15:15.638335       1 reflector.go:330] github.com/containous/traefik/vendor/k8s.io/client-go/1.5/tools/cache/reflector.go:109: watch of *v1.Service ended with: too old resource version: 137162 (138460)`
The messages are harmless and there is an issue for it:  https://github.com/kubernetes/kubernetes/issues/22024

### CODE REVIEW

**Comment by errm** [comment]
Wow awesome thanks @yvespp 

1. For now this sounds entirely reasonable... hopefully over time the k8s/client-go will get its dependencies under control.
2. I would like to avoid breaking backwards compat if we can, hopefully we can build a shim to handle the existing config, I am more than happy to deprecate the existing config with a warning as supporting the kube.config format sounds like a great idea.
3. Meh sounds fine, lets see how much support requests we get asking about this :)

I have taken a quick look at the code and it seems nice, but I am going to do a full review tomorrow....

**Comment by errm** [comment]
ping @containous/traefik

**Comment by yvespp** [comment]
I re-added the Endpoint configuration option so 2. is solved.

Adding the option to provide a kubeconfig file doesn't really make sense now. Traefik can only run inside the cluster because it forwards traffic directly to in cluster endpoints that are not normally reachable from outside.

File: provider/k8s/client.go [hunk 1]
```diff
@@ -1,296 +1,252 @@
...
+		c.clientset.ExtensionsClient,
+		"ingresses",
+		api.NamespaceAll,
+		fields.Everything(),
+		labelSelector)
+
+	c.ingStore, c.ingController = cache.NewInformer(
+		source,
+		&v1beta1.Ingress{},
+		0,
```
Comment (yvespp) [inline provider/k8s/client.go]: Probably a good idea, set it 3 minutes.

**Comment by emilevauge** [comment]
Hi @yvespp, thanks for you great work :)
The reason why we chose to create our own k8s client was that the official one [was huge](https://github.com/containous/traefik/issues/245#issuecomment-194244190). What is the size of the generated binary with this new client?
On you point 3, I assume this is the same reason that leads to issue https://github.com/containous/traefik/issues/532 (log every 60s ?).

**Comment by yvespp** [comment]
The traefik binary is 32M compiled via `make binary`.

To 3.: could be, we could dump the response so we would see what the server is actually sending. On my local cluster (minikube) I see the errors irregularly:

```
time="2016-11-13T20:40:15Z" level=info msg="Creating in cluster Kubernetes client" 
time="2016-11-13T20:40:15Z" level=info msg="Server configuration reloaded on :80" 
time="2016-11-13T20:40:15Z" level=info msg="Server configuration reloaded on :443" 
ERROR: logging before flag.Parse: W1113 20:55:04.388496       1 reflector.go:330] github.com/containous/traefik/vendor/k8s.io/client-go/1.5/tools/cache/reflector.go:109: watch of *v1.Service ended with: too old resource version: 185423 (186425)
ERROR: logging before flag.Parse: W1113 20:55:15.522916       1 reflector.go:330] github.com/containous/traefik/vendor/k8s.io/client-go/1.5/tools/cache/reflector.go:109: watch of *v1beta1.Ingress ended with: too old resource version: 185413 (186437)
ERROR: logging before flag.Parse: W1113 21:14:59.319868       1 reflector.go:330] github.com/containous/traefik/vendor/k8s.io/client-go/1.5/tools/cache/reflector.go:109: watch of *v1.Service ended with: too old resource version: 186426 (187734)
ERROR: logging before flag.Parse: W1113 21:15:16.533977       1 reflector.go:330] github.com/containous/traefik/vendor/k8s.io/client-go/1.5/tools/cache/reflector.go:109: watch of *v1beta1.Ingress ended with: too old resource version: 186438 (187753)
```

**Comment by emilevauge** [comment]
>32M

Ouch, the last release is 15MB https://github.com/containous/traefik/releases/tag/v1.1.0-rc4...

**Comment by errm** [comment]
This is large, but I am not sure its the end of the world. 

With compression the docker image comes out at 9.5mb vs 6.5mb for the previous version. Traefik will still be tiny compared to most things people deploy to containers...

**Comment by emilevauge** [comment]
Indeed, this is not the end of the world.
But if the official client still has connectivity issues with server (https://github.com/kubernetes/kubernetes/issues/22024, no update on this for several months), I wonder if it really worth it :'(
@yvespp Can you investigate a bit on those logs ?

**Comment by errm** [comment]
The main motivation for doing this is that we will reduce the load on the api server dramatically.  The current implementation makes many calls to the api server on each event we are watching.

**Comment by emilevauge** [comment]
@errm good point 
Nevertheless, @yvespp, if possible, could you investigate on connections dropped by server (rate?)?
I read somewhere that we could use websockets as a workaround, WDYT?

**Comment by errm** [comment]
I don't think that these error messages are an issue, so long as they are not happening often, the expected behaviour here is that we retrigger the watch...  I think the issue is something that is bubbling up from the etcd to the apiserver.  Besides, with these changes we will have parity with kubernetes internal services e.g. kubedns so we would expect traefik to be as resilent as them...

File: provider/k8s/client.go [hunk 2]
```diff
@@ -1,296 +1,274 @@
...
+	"k8s.io/client-go/1.5/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/1.5/pkg/fields"
+	"k8s.io/client-go/1.5/pkg/labels"
+	"k8s.io/client-go/1.5/pkg/runtime"
+	"k8s.io/client-go/1.5/pkg/watch"
+	"k8s.io/client-go/1.5/rest"
+	"k8s.io/client-go/1.5/tools/cache"
 )
 
+const resyncPeriod = time.Minute * 3
```
Comment (errm) [inline provider/k8s/client.go]: this looks good. We could go for 5 minutes to keep parity with what kubedns does https://github.com/kubernetes/kubernetes/blob/ff8e780c30884161700ce741743455f1afc2a02f/pkg/dns/dns.go#L55

**Comment by errm** [comment]
ping @yvespp we unfortunately need a rebase...

**Comment by yvespp** [comment]
Rebased and set resync to 5 minutes.

**Comment by yvespp** [comment]
Pro/con from my point of view:
* Pro: less code to maintain, less api request, robustness, easier to implement new features (e.g. certificates per host)
* Con: bigger binary, longer build time, dependency problem with golog

The bigger size is annoying but I think its worth it.
Should we open an issue for that?

**Comment by yvespp** [comment]
The websockts connection could still potentially drop and needs to be retried as well.
An easy way to avoid the problem would be to not use watches or websocket and just get all the needed resources every 2 seconds (or ProvidersThrottleDuration), as we currently can't change the config more frequently anyway. With this there would only be 3 API requests needed per ProvidersThrottleDuration.

The errors that are logged from the client (too old resource version) happen completely transparently and have no consequences. It seams to me this is expected behavior but just logged wrong (see https://github.com/kubernetes/kubernetes/issues/22024)

I analysed the dropping connections a bit more and it seems to be the same problem.
The connection in curl drops pretty much at the same time as the API Server logs this:
```
W1114 21:42:08.759556    1773 reflector.go:330] pkg/storage/cacher.go:194: watch of *extensions.Ingress ended with: 401: The event in requested index is outdated and cleared (the requested history has been cleared [198283/197927]) [199282]
W1114 21:42:09.760765    1773 cacher.go:469] Terminating all watchers from cacher *extensions.Ingress
```

Curl form inside a pod:
```
/var/run/secrets/kubernetes.io/serviceaccount # time curl --cacert ca.crt -H "Authorization: Bearer $(cat token)" 'https://10.0.0.1:443/apis/extensions/v1beta1/ingresses?watch=true' -v
*   Trying 10.0.0.1...
* TCP_NODELAY set
* Connected to 10.0.0.1 (10.0.0.1) port 443 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
ssl bla
> GET /apis/extensions/v1beta1/ingresses?watch=true HTTP/1.1
> Host: 10.0.0.1
> User-Agent: curl/7.51.0
> Accept: */*
> Authorization: Bearer eyJhbG....
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Mon, 14 Nov 2016 21:27:08 GMT
< Transfer-Encoding: chunked
< 
{"type":"ADDED","object":{"kind":"Ingress","apiVersion":"extensions/v1beta1","metadata":{"name":"echoserver","namespace":"default","selfLink":"/apis/extensions/v1beta1/namespaces/default/ingresses/echoserver","uid":"163ce154-a537-11e6-8eaf-42bf67fd163b","resourceVersion":"150437","generation":1,"creationTimestamp":"2016-11-07T22:11:09Z","labels":{"environment":"test"},"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"kind\":\"Ingress\",\"apiVersion\":\"extensions/v1beta1\",\"metadata\":{\"name\":\"echoserver\",\"creationTimestamp\":null},\"spec\":{\"rules\":[{\"http\":{\"paths\":[{\"path\":\"/echo\",\"backend\":{\"serviceName\":\"echoserver\",\"servicePort\":80}}]}}]},\"status\":{\"loadBalancer\":{}}}"}},"spec":{"rules":[{"http":{"paths":[{"path":"/echo","backend":{"serviceName":"echoserver","servicePort":80}}]}}]},"status":{"loadBalancer":{}}}}
{"type":"ADDED","object":{"kind":"Ingress","apiVersion":"extensions/v1beta1","metadata":{"name":"whoami-ingress","namespace":"default","selfLink":"/apis/extensions/v1beta1/namespaces/default/ingresses/whoami-ingress","uid":"6132361f-a9e1-11e6-b5a5-ae32eb2fa5e0","resourceVersion":"185413","generation":1,"creationTimestamp":"2016-11-13T20:40:14Z"},"spec":{"rules":[{"host":"foo.localhost","http":{"paths":[{"path":"/bar","backend":{"serviceName":"service1","servicePort":80}}]}}]},"status":{"loadBalancer":{}}}}
* Curl_http_done: called premature == 0
* Connection #0 to host 10.0.0.1 left intact
real	15m 1.48s
user	0m 0.00s
sys	0m 0.07s
```

**Comment by emilevauge** [comment]
@yvespp 
>The websockts connection could still potentially drop and needs to be retried as well.

In fact, I'm not worrying about connection drops due to network errors.
What bugs me is that long polling connections are deliberately dropped by k8s every 60s (by default, can be changed with [`min-request-timeout`](https://github.com/kubernetes/kubernetes/blob/release-1.2/docs/api.md)). I was wondering if using websocket instead would avoid this.

>An easy way to avoid the problem would be to not use watches or websocket and just get all the needed resources every 2 seconds (or ProvidersThrottleDuration), as we currently can't change the config more frequently anyway. With this there would only be 3 API requests needed per ProvidersThrottleDuration.

I don't think this would be better as we would have to recreate a connection very often. The point of using a long polling connection is to reuse an existing connection and also allow real time events.

>Pro: less code to maintain, less api request, robustness, easier to implement new features (e.g. certificates per host)
Con: bigger binary, longer build time, dependency problem with golog

On the dependency issue, a library [should not vendor its dependencies](https://github.com/kubernetes/client-go/issues/2), I don't get why they . I'm wondering why the [travis build](https://travis-ci.org/containous/traefik/builds/175832536) is green in fact ^^ Any idea?

**Comment by errm** [comment]
@yvespp I spent some time using glide to flatten the dependencies you can find my stuff here https://github.com/yvespp/traefik/compare/master...errm:flat-deps, If you could merge that into your branch I think that will cover the stuff @emilevauge raised

**Comment by yvespp** [comment]
I tried to rebase yesterday. Because of the merges in the glide.yaml I tryed to update the glide.lock via `glide up -v` but it fails with this error:

```
[INFO]	Found Godeps.json file in /root/.glide/cache/src/https-github.com-opencontainers-runc
[INFO]	--> Parsing Godeps metadata...
[INFO]	--> Fetching github.com/spf13/pflag.
[INFO]	--> Setting version for github.com/spf13/pflag to 08b1a584251b5b62f458943640fc8ebd4d50aaa5.
[INFO]	--> Fetching gopkg.in/inf.v0.
[INFO]	--> Fetching github.com/go-openapi/jsonpointer.
[INFO]	--> Fetching github.com/go-openapi/jsonreference.
[INFO]	--> Fetching github.com/go-openapi/swag.
[INFO]	--> Fetching github.com/emicklei/go-restful.
[INFO]	--> Setting version for github.com/emicklei/go-restful to 892402ba11a2e2fd5e1295dd633481f27365f14d.
[INFO]	--> Fetching github.com/juju/ratelimit.
[INFO]	--> Fetching github.com/pborman/uuid.
[INFO]	--> Fetching github.com/blang/semver.
[INFO]	--> Fetching github.com/PuerkitoBio/purell.
[INFO]	--> Fetching github.com/mailru/easyjson.
[INFO]	--> Fetching golang.org/x/oauth2.
[INFO]	--> Setting version for golang.org/x/oauth2 to 045497edb6234273d67dbc25da3f2ddbc4c4cacf.
[INFO]	--> Fetching github.com/coreos/go-oidc.
[INFO]	--> Fetching github.com/ghodss/yaml.
[INFO]	--> Setting version for github.com/ghodss/yaml to 73d445a93680fa1a78ae23a5839bad48f32ba1ee.
[ERROR]	Error scanning golang.org/x/net/http2: open /root/.glide/cache/src/https-golang.org-x-net/http2: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
```

Does anyone know how to fix this?
Should we add the instructions on how to update the glide.lock to the contributor guide? This should be needed every time a dependency in the glide.yaml changes?

**Comment by ekozan** [comment]
:/ linked to https://github.com/Masterminds/glide/issues/432

**Comment by errm** [comment]
I have merged the changes in from master and fixed the deps localy. I can give you a branch to grab my changes from, or alternatively I think github now has an option to allow maintainers to push to your branch...

**Comment by yvespp** [comment]
Whatever is more convenient for you. "Allow edits from maintainers" is checked so you should be able to commit to my master branch.

How did you fix it?

**Comment by errm** [comment]
Sorry I have messed up here, I ended up pushing my local master to your remote master, now github won't let me push because the PR has been closed... I have reopened a new PR #857, but feel free to reopen this one, if you can push the code back onto your branch.

**Comment by yvespp** [comment]
Ok, no problem. Glide is a mystery to me, I have no clue how I would have fixed the dependency problem, `glide up -v` still doesn't work on my machine...

If you want me to make more changes I will open a new pull request.
Can we merge this now? With your changes only 3. is open.

**Comment by errm** [comment]
@yvespp not sure about the glide issues you are seeing, the rebase on master broke everything for me, but `glide up -v` works fine now I have fixed...

@emilevauge WDYT about getting this merged soon, I am quite happy that this is an improvement, and we should be able to get any fixes needed done before 1.2?

File: provider/kubernetes.go [hunk 3]
```diff
@@ -123,17 +71,14 @@ func (provider *Kubernetes) Provide(configurationChan chan<- types.ConfigMessage
...
-						return err
 					case event := <-eventsChan:
 						log.Debugf("Received event from kubernetes %+v", event)
 						templateObjects, err := provider.loadIngresses(k8sClient)
 						if err != nil {
 							return err
 						}
 						if reflect.DeepEqual(provider.lastConfiguration.Get(), templateObjects) {
-							log.Debugf("Skipping event from kubernetes %+v", event)
+							log.Debug("Skipping event")
```
Comment (emilevauge) [inline provider/kubernetes.go]: Could you add `from kubernetes` to the message?

**Comment by yvespp** [comment]
@emilevauge Rebased, changed the debug log and also cleaned up the stop channels.

File: provider/k8s/client.go [hunk 4]
```diff
@@ -187,27 +187,21 @@ func (c *clientImpl) WatchEndpoints(stopCh <-chan struct{}) chan interface{} {
...
-	chanIngresses := c.WatchIngresses(kubeLabelSelector, stopIngresses)
-
-	stopServices := make(chan struct{})
-	chanServices := c.WatchServices(stopServices)
-
-	stopEndpoints := make(chan struct{})
-	chanEndpoints := c.WatchEndpoints(stopEndpoints)
+	chanIngresses := c.WatchIngresses(kubeLabelSelector, stopWatchCh)
+	chanServices := c.WatchServices(stopWatchCh)
+	chanEndpoints := c.WatchEndpoints(stopWatchCh)
```
Comment (emilevauge) [inline provider/k8s/client.go]: Using the same channel `stopWatchCh` in multiple go routines will fail :'(

**Comment by yvespp** [comment]
@emilevauge Hmm, I tested only using one channel specifically before changing it. Test App: https://play.golang.org/p/VDKmTeVTuj

Also the nginx ingress controller uses only one channel: https://github.com/kubernetes/contrib/blob/master/ingress/controllers/nginx/controller.go#L1225

File: provider/k8s/client.go (hunk 4 above)
Comment (emilevauge) [inline provider/k8s/client.go]: My bad, read too quickly ;)

File: build.Dockerfile [hunk 5]
```diff
@@ -21,6 +21,10 @@ WORKDIR /go/src/github.com/containous/traefik
 
 COPY glide.yaml glide.yaml
 COPY glide.lock glide.lock
-RUN glide install
+RUN glide install -v
+
+COPY integration/glide.yaml integration/glide.yaml
```
Comment (vdemeester) [inline build.Dockerfile]: :heart_eyes_cat:

**Comment by george-angel** [comment]
Eagerly awaiting this release, we have to use three `m4.large` instances for our masters to handle load from the traefik ingress proxies.

**Pull request merged**